### ISSUE TITLE
filesystem HMD_IMAGE_UPLOAD_TYPE and restart

### ISSUE THREAD

**Issue by orthagh** [hackmdio/docker-hackmd#15]
Hello.

i encounter an issue with my docker instance.
I'm currently using "filesystem" value for ```HMD_IMAGE_UPLOAD_TYPE``` so the upload files are stored in docker fs.

On config.json changes and after a docker-compose up, all my uploads vanish :cry: 

I retrieved them from a vm backup, but do you some hints to avoid this behavior ?

**Comment by SISheogorath** [comment]
Sorry to hear.

When I build the warning I didn't think of people who edit the config file as the normal way to configure containers is using environment variables. If you use them to configure a file upload you get a big warning if you don't configure a volume.

https://github.com/hackmdio/docker-hackmd/blob/master/hackmd/docker-entrypoint.sh#L8-L22

We don't define a volume by default because it's not needed and produces empty unnamed volumes when you don't use compose.

Once the repository restructuring is merged and completed, you shouldn't need to build the image yourself.

Maybe you can explain, why you preferred editing the config.json over the usage of the environment variables :)

**Comment by orthagh** [comment]
Sorry, it's a mistake in my previous post.
I edit ```docker-compose.yml``` file not config.json.

Here is the content
```yml
hackmdPostgres:
  image: postgres
  environment:
    - POSTGRES_USER=hackmd
    - POSTGRES_PASSWORD=hackmdpass
    - POSTGRES_DB=hackmd
  restart: always
hackmd:
  image: hackmdio/hackmd:latest
  environment:
    - POSTGRES_PASSWORD=hackmdpass
    - HMD_DOMAIN=...
    - HMD_PORT=80
    - HMD_URL_ADDPORT=false
    - HMD_ALLOW_FREEURL=true
    - HMD_USECDN=false
    - HMD_ALLOW_EMAIL_REGISTER=false
    - HMD_EMAIL=false
    - HMD_ALLOW_ANONYMOUS=false
    - HMD_IMAGE_UPLOAD_TYPE=filesystem
    - HMD_GITLAB_BASEURL=...
    - HMD_GITLAB_CLIENTID=...
    - HMD_GITLAB_CLIENTSECRET=...
    - HMD_LDAP_URL=...
    - HMD_LDAP_SEARCHBASE=...
    - HMD_LDAP_SEARCHFILTER=...
    - HMD_LDAP_PROVIDERNAME=...
  links:
    - hackmdPostgres:hackmdPostgres
  ports:
    - "80:80"
  restart: always
```

**Comment by SISheogorath** [comment]
In this case, there was a warning printed to your log, that you loose your data if you don't specify a volume.

In general, it's normal that docker containers don't persist data as long as you do not specify a volume. Changing something in the compose file recreates the (possibly) affected containers. In other words, it removes the old ones and creates new with the new config. Everything that is no in a volume vanishes.

You **can** define volumes in the Dockerfile. But you don't have to. The default settings are made to not store any local data for reason. That's why we don't define a local volume because it only pollutes setup. If you change default settings you are responsible for their effects.

Currently, the documentation is not the best, so yes, mentioning it in the README is a good idea and should be done soon. 

And you are lucky. Because the only reason you not loose your Postgres database is that the Dockerfile specifies a volume and docker-compose handles them itself. You should specify a volume for Postgres as well and migrate your data.

**Comment by orthagh** [comment]
yes, i saw that for postgres.
I'm currently doing the volume stuff to mount the /hackmd/public/upload folder from outside the container but my though was it could be great to enhance the default readme.

I just saw the warning (i'm using -d option), so thank you :)

**Comment by johanneslotz** [comment]
for the record: on my installation the uploads are at /hackmd/public/upload**s**, so i am adding

`
volumes:
    - ./uploads:/hackmd/public/uploads
`

**Comment by SISheogorath** [comment]
If there are any further issues or request, feel free to re-open or open a new issue
